File: macos/Sources/PrayCalcCore/qibla.c
/**
 * qibla.c — Qibla bearing and distance calculation.
 *
 * Port of web/lib/qibla.ts. Uses the forward azimuth (initial bearing)
 * formula on the great circle to the Ka'bah center coordinates.
 */

#include "qibla.h"
#include <math.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/* Exact center of the Ka'bah, Masjid al-Haram, Mecca */
#define KAABA_LAT 21.422511
#define KAABA_LNG 39.826150

static double to_rad(double d) { return d * M_PI / 180.0; }

double qibla_bearing(double lat, double lng) {
    double phi1 = to_rad(lat);
    double phi2 = to_rad(KAABA_LAT);
    double d_lambda = to_rad(KAABA_LNG - lng);

    double y = sin(d_lambda) * cos(phi2);
    double x = cos(phi1) * sin(phi2) - sin(phi1) * cos(phi2) * cos(d_lambda);

    double bearing = atan2(y, x) * 180.0 / M_PI;
    return fmod(bearing + 360.0, 360.0);
}

double qibla_distance_km(double lat1, double lng1, double lat2, double lng2) {
    const double R = 6371.0;
    double d_lat = to_rad(lat2 - lat1);
    double d_lng = to_rad(lng2 - lng1);
    double a = sin(d_lat / 2) * sin(d_lat / 2) +
               cos(to_rad(lat1)) * cos(to_rad(lat2)) *
               sin(d_lng / 2) * sin(d_lng / 2);
    return R * 2.0 * atan2(sqrt(a), sqrt(1.0 - a));
}

File: web/src/lib/qibla.ts
/**
 * qibla.ts — Qibla direction utilities.
 *
 * PURPOSE: Pure math for Qibla bearing calculation.
 *   Safe for client islands (no external deps).
 * INPUTS: lat/lng coordinates
 * OUTPUTS: bearing in degrees, compass abbreviation
 * CONSTRAINTS: No server-only imports
 * REF: P2-E3-W02-S02-T03
 */

/** Exact center of the Ka'bah, Masjid al-Haram, Mecca */
export const KAABA_LAT = 21.422511;
export const KAABA_LNG = 39.826150;

const toRad = (d: number) => (d * Math.PI) / 180;

/** Qibla bearing in degrees clockwise from north (0°=N, 90°=E, 180°=S, 270°=W) */
export function qiblaAngle(lat: number, lng: number): number {
  const φ1 = toRad(lat), λ1 = toRad(lng);
  const φ2 = toRad(KAABA_LAT), λ2 = toRad(KAABA_LNG);
  const y = Math.sin(λ2 - λ1) * Math.cos(φ2);
  const x =
    Math.cos(φ1) * Math.sin(φ2) -
    Math.sin(φ1) * Math.cos(φ2) * Math.cos(λ2 - λ1);
  return ((Math.atan2(y, x) * 180) / Math.PI + 360) % 360;
}

/** 8-point compass abbreviation for a bearing */
export function compassDir(bearing: number): string {
  const dirs = ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW'];
  return dirs[Math.round(bearing / 45) % 8] ?? 'N';
}
